File: python/packages/kagent-adk/src/kagent/adk/_mcp_apps.py
"""MCP App (UI) tool result compaction for the model.

Mirrors the Go ADK behavior in ``go/adk/pkg/agent/mcp_apps.go``. An MCP App tool
(one declaring a ``ui.resourceUri``) renders an interactive widget in the chat
that updates itself in place. The model must treat a successful render as a
completed, self-updating artifact; otherwise it tends to re-invoke the rendering
tool on every "refresh", flooding the chat with duplicate app cards (observed
with weaker models calling the same render tool 5-7 times in a row).

We keep the full tool payload in the session/chat history so the UI can render
the widget, and only rewrite what the *model* sees into a short terminal
directive. ADK builds the model request from deep copies of the session events
(see ``flows/llm_flows/contents.py``), so mutating the request here does not
corrupt the stored history.
"""

from __future__ import annotations

from google.adk.agents.callback_context import CallbackContext
from google.adk.models.llm_request import LlmRequest

# Terminal directive the model sees in place of an MCP App tool's render
# payload. It is protocol-oriented: it applies to any tool carrying a UI
# resourceUri, independent of the tool's name or payload keys.
MCP_APP_RENDERED_NOTICE = (
    "The interactive UI for this tool has been rendered to the user and now "
    "updates live inside the app. Treat this as complete and do not call this "
    "tool again unless the user explicitly asks for it."
)


class MCPAppToolNames:
    """Mutable, shared set of MCP App (UI-rendering) tool names.

    Populated lazily by ``KAgentMcpToolset.get_tools`` as MCP tools are resolved
    (which happens during request preprocessing, before ``before_model_callback``
    runs) and read by the compaction callback. Using a shared object avoids
    re-listing MCP tools on every model turn.
    """

    def __init__(self) -> None:
        self._names: set[str] = set()

    def add(self, name: str) -> None:
        self._names.add(name)

    def __contains__(self, name: str) -> bool:
        return name in self._names

    def __bool__(self) -> bool:
        return bool(self._names)

    def __len__(self) -> int:
        return len(self._names)


def compact_mcp_app_response(response: dict) -> dict:
    """Rewrite an MCP App tool result (a JSON ``CallToolResult``) for the model.

    On error, keep the content so the model can diagnose/recover but drop the
    heavy structured payload. On success, collapse the render payload into a
    terminal directive so the model stops re-invoking the rendering tool,
    preserving ``_meta`` (e.g. resourceUri) for any downstream tooling.
    """
    if response.get("isError") is True or response.get("error") is True:
        compacted = dict(response)
        compacted.pop("structuredContent", None)
        return compacted

    compacted: dict = {"content": [{"type": "text", "text": MCP_APP_RENDERED_NOTICE}]}
    if "_meta" in response:
        compacted["_meta"] = response["_meta"]
    return compacted


def make_mcp_app_model_result_callback(app_tool_names: MCPAppToolNames):
    """Build a ``before_model_callback`` that compacts MCP App tool results.

    Only the model's view is changed; the full result remains in chat history
    for UI rendering. The callback is a no-op until ``app_tool_names`` has been
    populated, so it is safe to attach unconditionally.
    """

    def before_model(callback_context: CallbackContext, llm_request: LlmRequest) -> None:
        if not app_tool_names or not llm_request.contents:
            return None
        for content in llm_request.contents:
            for part in content.parts or []:
                function_response = part.function_response
                if function_response is None or function_response.name not in app_tool_names:
                    continue
                if isinstance(function_response.response, dict):
                    function_response.response = compact_mcp_app_response(function_response.response)
        return None

    return before_model

File: go/adk/pkg/agent/mcp_apps.go
package agent

import (
	"encoding/json"

	mcpsdk "github.com/modelcontextprotocol/go-sdk/mcp"
	"google.golang.org/adk/v2/agent"
	"google.golang.org/adk/v2/agent/llmagent"
	adkmodel "google.golang.org/adk/v2/model"
)

// mcpAppRenderedNotice is the terminal message the model sees in place of an
// MCP App tool's render payload. An MCP App tool (one that declares a UI
// resourceUri) produces an interactive view that is displayed to the user and
// refreshes itself in-place via its own app-only tool calls. The model must
// treat a successful render as a completed, self-updating artifact; otherwise it
// tends to re-invoke the rendering tool on every "refresh", flooding the chat
// with duplicate app cards. This notice is protocol-oriented: it applies to any
// tool carrying a UI resourceUri, independent of the tool's name or payload keys.
const mcpAppRenderedNotice = "The interactive UI for this tool has been rendered to the user and now updates live inside the app. Treat this as complete and do not call this tool again unless the user explicitly asks for it."

// MakeMCPAppModelResultCallback replaces what the model sees for MCP App
// (UI-rendering) tool results: instead of the heavy render payload it receives a
// terminal directive (see mcpAppRenderedNotice). The full result is still
// streamed to the UI separately, so this only changes the model's view and
// prevents the model from looping on the rendering tool. Errors are passed
// through so the model can still react to and recover from failures.
func MakeMCPAppModelResultCallback(appToolNames map[string]bool) llmagent.BeforeModelCallback {
	return func(_ agent.Context, req *adkmodel.LLMRequest) (*adkmodel.LLMResponse, error) {
		for _, content := range req.Contents {
			if content == nil {
				continue
			}
			for _, part := range content.Parts {
				if part == nil || part.FunctionResponse == nil || !appToolNames[part.FunctionResponse.Name] {
					continue
				}
				part.FunctionResponse.Response = compactMCPAppModelResponse(part.FunctionResponse.Response)
			}
		}
		return nil, nil
	}
}

// compactMCPAppModelResponse rewrites an MCP App tool result for the model.
//
// The model exchanges tool results as a generic map (genai
// FunctionResponse.Response), but the payload is really an MCP
// [mcpsdk.CallToolResult]. We decode it into that typed result so the logic
// works against real fields (IsError, Content, Meta, StructuredContent) rather
// than poking at string keys. If the payload isn't a recognizable MCP result we
// leave it untouched.
func compactMCPAppModelResponse(response map[string]any) map[string]any {
	result, err := decodeCallToolResult(response)
	if err != nil {
		return response
	}

	if result.IsError {
		// On error, keep the original content/meta so the model can
		// diagnose and recover; only drop the heavy structured payload.
		result.StructuredContent = nil
		return encodeCallToolResult(result, response)
	}

	// On success, collapse the render payload into a terminal directive so the
	// model stops re-invoking the rendering tool. Preserve _meta (e.g.
	// resourceUri) in case downstream tooling relies on it.
	compact := &mcpsdk.CallToolResult{
		Meta:    result.Meta,
		Content: []mcpsdk.Content{&mcpsdk.TextContent{Text: mcpAppRenderedNotice}},
	}
	return encodeCallToolResult(compact, response)
}

// decodeCallToolResult interprets a generic model-facing response map as a typed
// MCP CallToolResult.
func decodeCallToolResult(response map[string]any) (*mcpsdk.CallToolResult, error) {
	raw, err := json.Marshal(response)
	if err != nil {
		return nil, err
	}
	var result mcpsdk.CallToolResult
	if err := json.Unmarshal(raw, &result); err != nil {
		return nil, err
	}
	return &result, nil
}

// encodeCallToolResult converts a typed CallToolResult back into the generic map
// the model expects, falling back to the original response if conversion fails.
func encodeCallToolResult(result *mcpsdk.CallToolResult, fallback map[string]any) map[string]any {
	raw, err := json.Marshal(result)
	if err != nil {
		return fallback
	}
	var out map[string]any
	if err := json.Unmarshal(raw, &out); err != nil {
		return fallback
	}
	return out
}
